go/kbfs/simplefs/downloads.go

```diff
@@ -0,0 +1,353 @@
+		d.state.LocalPath = localPath
+		return d
+	}); err != nil {
+		done(err)
+		return
+	}
+	done(nil)
+}
+
+func (m *downloadManager) startDownload(
+	ctx context.Context, arg keybase1.SimpleFSStartDownloadArg) (
+	downloadID string, err error) {
+	opid, err := m.k.SimpleFSMakeOpid(ctx)
+	if err != nil {
+		return "", err
+	}
+	downloadID = strconv.FormatInt(time.Now().UnixNano(), 16)
+	downloadPath, filename, err := m.getDownloadPath(ctx, arg.Path, downloadID)
+	if err != nil {
+		return "", err
+	}
+	// TODO for dirs maybe we want zip instead?
+	err = m.k.SimpleFSCopyRecursive(ctx, keybase1.SimpleFSCopyRecursiveArg{
+		OpID: opid,
+		Src:  keybase1.NewPathWithKbfs(arg.Path),
+		Dest: keybase1.NewPathWithLocal(downloadPath),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	func() {
+		defer m.publisher.DownloadStatusChanged()
+		m.lock.Lock()
+		defer m.lock.Unlock()
+		m.downloads[downloadID] = download{
+			info: keybase1.DownloadInfo{
+				DownloadID:        downloadID,
+				Path:              arg.Path,
+				Filename:          filename,
+				StartTime:         keybase1.ToTime(time.Now()),
+				IsRegularDownload: arg.IsRegularDownload,
+			},
+			opid: opid,
+			state: keybase1.DownloadState{
+				DownloadID: downloadID,
+			},
+		}
+	}()
+
+	bgCtx, cancelBtCtx := m.makeContext()
+	done := func(err error) {
+		_ = m.updateDownload(downloadID, func(d download) download {
+			if d.state.Done || d.state.Canceled || len(d.state.Error) > 0 {
+				return d
+			}
+			if err == context.Canceled {
+				d.state.Canceled = true
+			} else if err != nil {
+				d.state.Error = err.Error()
+			} else {
+				d.state.EndEstimate = keybase1.ToTime(time.Now())
+				d.state.Progress = 1
+				d.state.Done = true
+			}
+			return d
+		})
+		cancelBtCtx()
+	}
+	go m.monitorDownload(bgCtx, opid, downloadID, done)
+	go m.waitForDownload(bgCtx, downloadID, downloadPath, done)
+
+	return downloadID, nil
+}
+
+func (m *downloadManager) getDownloadStatus(ctx context.Context) (
+	status keybase1.DownloadStatus) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, download := range m.downloads {
+		status.States = append(status.States, download.state)
+		if download.info.IsRegularDownload {
+			status.RegularDownloadIDs = append(
+				status.RegularDownloadIDs, download.info.DownloadID)
+		}
+	}
+	sort.Slice(status.RegularDownloadIDs, func(i, j int) bool {
+		d1, ok := m.downloads[status.RegularDownloadIDs[i]]
+		if !ok {
+			return false
+		}
+		d2, ok := m.downloads[status.RegularDownloadIDs[j]]
+		if !ok {
+			return false
+		}
+		return d1.info.StartTime.After(d2.info.StartTime)
+	})
+	return status
+}
+
+func (m *downloadManager) cancelDownload(
+	ctx context.Context, downloadID string) error {
+	d, err := m.getDownload(downloadID)
+	if err != nil {
+		return err
+	}
+	return m.k.SimpleFSCancel(ctx, d.opid)
+}
+
+func (m *downloadManager) dismissDownload(
+	ctx context.Context, downloadID string) {
+	m.cancelDownload(ctx, downloadID)
```

Comment:
Not checking the error here.

Reply:
it's possible the download entry is already removed by now; this just makes sure it's canceled in case caller didn't call cancel explicitly. I'll add a comment.

Reply:
I meant that the linter is going to yell at you.  You have to at least do `_ =` here.  Though maybe `cancelDownload` shouldn't throw an error if it's already canceled, if that's an expected condition?

Reply:
it actually doesn't error if it's already canceled, but errors if a download is not found (i.e. dismissed). I'm gonna clarify in comment.